### ISSUE TITLE
Backward compatibility (StackTrace)

### ISSUE THREAD

**Issue by lpgauth** [tomas-abrahamsson/gpb#154]
Hi, it seems that the code generated is not compatible with < OTP20. The last two blocks should be inverted?

```erlang
-ifdef('OTP_RELEASE').
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason:StackTrace -> error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-else.
-ifdef('GPB_FUNCTION_STACK').
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason ->
        StackTrace = erlang:get_stacktrace(),
        error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-else.
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason:StackTrace -> error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-endif.

-endif.
```

**Comment by tomas-abrahamsson** [comment]
Was that code generated on Erlang 21? On Erlang 20, the code below is generated. As you can see, it defaults to work on the Erlang version used to generate it, but with an override. There's also an option, `{target_erlang_version, integer() | current}` for targetting a version.

```erlang
-ifdef('OTP_RELEASE').
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason:StackTrace -> error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-else.
-ifdef('GPB_PATTERN_STACK').
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason:StackTrace -> error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-else.
decode_msg_1_catch(Bin, MsgName, TrUserData) ->
    try decode_msg_2_doit(MsgName, Bin, TrUserData)
    catch Class:Reason ->
        StackTrace = erlang:get_stacktrace(),
        error({gpb_error,{decoding_failure, {Bin, MsgName, {Class, Reason, StackTrace}}}})
    end.
-endif.

-endif.
```

**Comment by tomas-abrahamsson** [comment]
I'd like to draw attention to the fact that the `-ifdef` override was meant as a temporary measure, serving a purpose during the 21 release candidates, before the `OTP_RELEASE` macro was introduced, see [this comment in `gpb_gen_decoders.erl`](https://github.com/tomas-abrahamsson/gpb/blob/4.3.3/src/gpb_gen_decoders.erl#L122-L131).

My intention is to remove the `GPB_PATTERN_STACK` and `GPB_FUNCTION_STACK` overrides soon.

**Comment by lpgauth** [comment]
Shouldn't the code generated be backward / forward compatible regardless of the OTP version used to generate the code? In this case the code was generated on OTP-21 and doesn't work on OTP-20.

**Comment by tomas-abrahamsson** [comment]
Yes, you're right, I see your point now. Sorry. All the more reason to remove the inner `GPB_PATTERN_STACK` and `GPB_FUNCTION_STACK` checks that makes it all to easy to meander wrongly when reasoning about the code, and let the `OTP_RELEASE` check do all the work now that it exists. I will fix (and try to add some cross-version targetting checks to my local testing machinery)

**Comment by tomas-abrahamsson** [comment]
> Shouldn't the code generated be backward / forward compatible regardless of the OTP version used to generate the code? In this case the code was generated on OTP-21 and doesn't work on OTP-20.

I'd like to revisit this: No, in the general case, the above is not possible, hence the `target_erlang_version` option. For Erlang 21 it is possible due to the `OTP_RELEASE` macro, which was added in Erlang 21. But for instance between Erlang 18 and Erlang 19, it is not possible[*]. In Erlang 19, it became possible in type specs to indicate that a map key must exist using `:=` or may exist using `=>` which is of relevance for gpb. There can sometimes be a need to target earlier versions, for instance [`hex_core` is targeting Erlang 17](https://github.com/hexpm/hex_core/blob/v0.2.0/rebar.config#L11)

When using the `target_erlang_version` option, the generated code compiles without warnings, both when targeting an earlier or later version (I've checked). I will still simplify the generated code by removing the `GPB_PATTERN_STACK` and `GPB_FUNCTION_STACK` ifdef checks, but I figure there's no pressing for a bugfix version, due to the option, so I'll just include it in the next release whenever that'll be available.

____
[*] at least not without involving the build system, but that would be a bit outside the scope of the generated code

**Comment by lpgauth** [comment]
I'm a bit confused since in this case it's easy to be backward and forward compatible...

https://github.com/lpgauth/rig/commit/fe3c83e403a0ac5f8a205c47bd05b2de96f77395

Anyways, I'll add a note to our repo to generate the code using OTP-20 for now.
